alerting/pagerduty.go

```diff
@@ -0,0 +1,105 @@
+package alerting
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/datajet-io/peekaboo/retry"
+	"github.com/datajet-io/peekaboo/services"
+	"github.com/uber-go/zap"
+)
+
+const apiEndpoint = "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
+
+//Pagerduty represents the configuration of the Pageruty account used for notifications
+type PagerdutyClient struct {
+	integrationKey string
+}
+
+type PagerdutyEvent struct {
+	Service_key  string                 `json:"service_key"`
+	Incident_key string                 `json:"incident_key"`
+	Event_type   string                 `json:"event_type"`
+	Description  string                 `json:"description"`
+	Client       string                 `json:"client"`
+	Details      map[string]interface{} `json:"details"`
+}
+
+func NewPagerdutyClient(key string) *PagerdutyClient {
+	return &PagerdutyClient{integrationKey: key}
+}
+
+func (p PagerdutyClient) Print() {
+	fmt.Printf("P IntegrationKey is: %s\n", p.integrationKey)
+}
+
+func (p PagerdutyClient) TriggerAlert(service services.Service, alert Alert, details map[string]interface{}) error {
+	eventData := &PagerdutyEvent{
+		Service_key:  p.integrationKey,
+		Incident_key: service.Name,
+		Event_type:   "trigger",
+		Description:  alert.Message,
+		Client:       "peekaboo",
+		Details:      details,
+	}
+
+	eventDataJson, err := json.Marshal(eventData)
+	if err != nil {
+		service.Logger.Warn(
+			"Error marshaling 'eventData' to JSON",
+			zap.Object("eventData", eventData),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	// Make request, with retrying support
+	operation := func() error {
+		req, err := http.NewRequest("POST", apiEndpoint, bytes.NewBuffer(eventDataJson))
```

Comment:
DONE

Couldn't return since we're in a retrying block, but I moved these setup tasks outside the retried section, so it will immediately fail in that case and skip the retries. Good catch.